Add support for custom mermaid.initialize() options.

jaxbulsara opened the pull request:
I wanted to update the theme of my mermaid charts, but that required a custom argument to `mermaid.initialize()`.

I added a configuration options `mermaid_init` to be added to `conf.py` that can hold a dictionary of options to be passed to `mermaid.initialize`. When the JS files are being added to the app, I check if the configuration option is set. If it is, I use `json.dumps()` to convert the dictionary to a JSON string compatible with JavaScript. Otherwise, I use the default initialization string.

One pitfall with this approach is that there is no verification that the values within `mermaid_init` are compatible in JavaScript after the JSON dump. 

Additionally, I updated the build process to use a VERSION file that can be version controlled itself instead of an environment variable. This way, the version associated with any release can be verified within the source code.

jaxbulsara committed:
Add configuration option to specify parameters to .

jaxbulsara committed:
Update config option formatting.

jaxbulsara committed:
Add link to MermaidJS configuration docs.

jaxbulsara committed:
Rename to indicate the default value.

jaxbulsara committed:
Check if mermaid_init type is valid. Raise error otherwise.

jaxbulsara committed:
Add type hint.

jaxbulsara commented:
I looked back and saw why the VERSION was changed to being an environment variable. I think I will remove my changes to the build process so deployments via github actions don't break.

Doreapp commented on README.md:
```diff
@@ -23,6 +23,29 @@ extensions = [
+
+``mermaid_init``
+
+Set to a dictionary of values to pass to `mermaid.initialize()`. Find more info
```
Maybe put this on a single line ?

```suggestion
`mermaid_init` can be set as a dictionary of values. 
These will be passed to `mermaid.initialize()`. Find more info
```

Doreapp commented on README.md:
````diff
@@ -23,6 +23,29 @@ extensions = [
+```python
+mermaid_init = {
+  'theme': 'base',
+  'themeOptions': {
````
According to [the doc](https://mermaid.js.org/config/theming.html#customizing-themes-with-themevariables), it is rather `themeVariables`

```suggestion
  'themeVariables': {
```

jaxbulsara replied:
That's so weird because I swear I copy-pasted the example on that page. But I definitely see "themeVariables" in the docs.

Doreapp commented on sphinxmermaid/__init__.py:
```diff
@@ -68,13 +69,42 @@ def install_js(app: "Sphinx", *_):
+    """
+    Returns the `mermaid.initialize({...})` code string from the value
+    specified in conf.py or the default value.
+    """
+
+    mermaid_init = app.config.mermaid_init
+
+    if mermaid_init:
+        check_mermaid_init(mermaid_init)
+
+        params = json.dumps(mermaid_init)
+        return f"mermaid.initialize({params})"
+
+    return DEFAULT_MERMAID_INIT
```
I am not a big fan of all these line breaks... What do you think of removing them ? (Open to discussion)

jaxbulsara replied:
It does look out of place in the style of the rest of the code. I write really spacey code for legibility, but consistency matters more. I'll go ahead and remove them.

Doreapp commented:
Can you remove the reverted commits (as well as the commits made to revert a commit) from the history ?

Can be done by dropping the commits during an interactive rebase, example:

```bash
git rebase -i fd15b07e6599368ecbd5f5ab30d51de525622131
```

```
pick 86e2fe3 Add configuration option to specify parameters to .
pick 1cee0ae Update config option formatting.
pick 435232b Add link to MermaidJS configuration docs.
d df9218a Use VERSION file to hold package version.
pick 9d09ded Rename to indicate the default value.
pick dd80d8d Check if mermaid_init type is valid. Raise error otherwise.
pick 8ec71c4 Add type hint.
d e8e3c3a Increment version.
d f98b1a8 Revert "Increment version."
d bc4bcd7 Revert "Use VERSION file to hold package version."
```

Doreapp reviewed (changes requested):
LGTM! 👌 

Good feature added, and nicely done!

Doreapp commented on sphinxmermaid/__init__.py:
```diff
@@ -12,7 +13,7 @@
 LOGGER = logging.getLogger(__name__)
 MERMAID_JS_URL = "https://unpkg.com/mermaid/dist/mermaid.min.js"
-JS_INIT_MERMAID = "mermaid.initialize({startOnLoad:true});"
+DEFAULT_MERMAID_INIT = "mermaid.initialize({startOnLoad:true});"
```
This may be defined as the default parameters dictionary, and not as a raw javascript instruction, don't you think ?

It could then become:

```python
DEFAULT_MERMAID_INIT = {"startOnLoad": True}

# ...

def create_mermaid_init(app: "Sphinx"):
    mermaid_init = app.config.mermaid_init
    if mermaid_init:
        check_mermaid_init(mermaid_init)
    else:
        mermaid_init = {}
    mermaid_init.update(DEFAULT_MERMAID_INIT)
    params = json.dumps(mermaid_init)
    return f"mermaid.initialize({params})"
```

jaxbulsara replied:
I agree. See my commit below for the update.

Instead of updating the user-specified mermaid_init with the default values, I believe that the user should be able to completely override the default dictionary. In my version, the default is only used if `sphinxmermaid_mermaid_init` is not set.

Doreapp replied:
All right, that's fine with me.

jaxbulsara committed:
Reformat mermaid_init config definition.

Co-authored-by: AntoineMdn <55652808+Doreapp@users.noreply.github.com>

jaxbulsara committed:
Fix themeVariables key.

Co-authored-by: AntoineMdn <55652808+Doreapp@users.noreply.github.com>

jaxbulsara committed:
Compacted code for consistent style.
Moved create_mermaid_init() to add_js_file().

jaxbulsara committed:
Change default mermaid init to native dict.
Explicitly check if configured mermaid init is None to choose between user config and default.
User should be able to completely override default options.

jaxbulsara committed:
Update configuration option name to have extension name as prefix.
To ensure uniqueness.

jaxbulsara committed:
Merge upstream changes with config option name.

jaxbulsara committed:
Update config option name in error message.

Doreapp commented:
@jaxbulsara 

Note: I release a new version with your changes included, [v0.0.8](https://github.com/Doreapp/sphinx-mermaid/releases/tag/0.0.8)
